### DISCUSSION TITLE
has anyone thought of turning mithril-js into lisp?

### DISCUSSION THREAD

**Discussion by ghost** [MithrilJS/mithril.js#2800] [Q&A] [answered]
Hi! 

*this discussion here is out of technical curiosity - I really like mithril-js and lisp, I would like to know if you all already thought about this possible integration.* 

I would like to program mithril-js as lisp, would that be possible?... how to turn mithril-js like as lisp?... has anyone thought of turning mithril-js into lisp?

#### code1_with_mithril_js.html
```html
<html>
<body>
    <script src="https://unpkg.com/mithril/mithril.js"></script>
    <script>
    var root = document.body
    m.render(root, [
    m("main", [
        m("h1", {class: "title"}, "My first app"),
        m("button", "A button"),
         ])
      ])
    </script>
</body>
</html>
```

#### code1_lisp_with_mithril_js.html
```html
<html>
<body>
    <script src="https://unpkg.com/mithril/mithril.js"></script>
    <script>
    var root = document.body
    render(root, (
      ("main", (
         ("h1", (class: "title"), "My first app"),
         ("button", "A button"),
        ))
     ))
    </script>
</body>
</html>
```

**Comment by pygy** [comment, marked as answer]
This wouldn't work in vanilla JS, because

```js
("button", "A button")
```

evaluates to `"A button"` because of the coma operator semantics.

We could use arrays `["button", "A button"]`, but it wouldn't be without problems both semantically and perf-wise. In Mithril, `m('a', 'b')` is a `<a>` virtual element, whereas ,`['a', 'b']` is a fragment of two strings. So we'd need a special syntax for fragments.

The `m()` factory also returns objects that have a static structure, which lets the mithril core deal with it efficiently (we say it's monomorphic code, vs. polymorphic). This lets JS engines rely on the hidden class to access the various fields efficiently, making diff and render faster.

**Comment by panoply** [comment]
I explored this concept briefly using a build step (though not lisp). The goal was bring auto-completion of tags and in addition css/scss class names. It was rather pleasant to write an app with, eg:

```js
m.div(
  'row',
  'jc-center'
)(
  m.div(
    'col',
    'col-sm-6',
    'col-lg-12'
  )(
    m.ul('list')([
      m.li('list-item')('one'),
      m.li('list-item')('two')
    ])
  )
)
```
I've considered porting into a esbuild plugin but meh. 

---


https://user-images.githubusercontent.com/7041324/189385155-69bb60f8-27a4-43e7-9030-14d8394a274d.mp4
